Feature/Add docker setup for dev and prod

Philipp36 opened the pull request:
- Dockerfile with WSGI-server and odbc drivers
- Dockerfile with mssql tooling
- Examples for docker-compose stacks (for dev and prod)
- Rework main to be able to create app objects used by wsgi server
- Documentation how to use docker images

Philipp36 committed:
Feature/Add docker setup for dev and prod

- Dockerfile with WSGI-server and odbc drivers
- Dockerfile with mssql tooling
- Examples for docker-compose stacks (for dev and prod)
- Rework main to be able to create app objects used by wsgi server
- Documentation how to use docker images

slavistan commented on main.py:
```diff
@@ -39,10 +41,16 @@ def make_conn():
 if __name__ == "__main__":
-    main()
+    debug = os.getenv("DEBUG") is not None
+    webserver_hostname = os.getenv("WEBSERVER_HOSTNAME", "localhost")
```
Würde eher "dashboard" als Präfix verwenden anstatt dem etwas abstrakten "webserver", oder?

Philipp36 replied:
Ja, ist vermutlich klarer.

slavistan commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+  db:
+    image: mcr.microsoft.com/mssql/server:2022-latest
+    ports:
+      - "1433:1433/tcp"
```
Hier ist der Port noch hardgecoded, nutzt die MSSQL_PORT Envvar nicht. Ich würde den (Host-)Port gerne konfigurieren können.

slavistan commented on docker/development/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+        && rm -rf /var/lib/apt/lists/*
+
+# Make executables of sqltools and sqlpackage available in PATH
+ENV PATH="${PATH}:${SQLPACKAGE_FOLDER}:/opt/mssql-tools18/bin"
```
Leg die Binary unter (/usr)/bin ab, dann musst du nicht am PATH herumoperieren.

slavistan commented on docker/development/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+
+RUN unzip ${SQLPACKAGE_FILE_NAME} -d ${SQLPACKAGE_FOLDER}
+
+RUN chmod a+x ${SQLPACKAGE_FOLDER}/sqlpackage
```
All diese Instruktionen gehören zusammen, lösch doch bitte die Leerzeilen.

slavistan commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    depends_on:
+      db:
+        condition: service_healthy
+    command: /db-init/load_bacpac.sh
```
Nur aus Interesse: Weshalb definierst du das Startkommando _hier_ anstatt direkt im Dockerfile als ENTRYPOINT/CMD?

Philipp36 replied:
Das Dockerfile unter _docker/development/Dockerfile_ bzw. jetzt unter _dev/Dockerfile_ ist mehr oder weniger generisch (Installiert nur Abhängigkeiten, aber kann grundsätzlich für verschiedene Dinge verwendet werden).
Leider bricht die Idee gerade an ein paar Stellen, weshalb es auch nicht offensichtlich ist, wie es gedacht ist.

slavistan commented on README.md:
````diff
@@ -0,0 +1,47 @@
+The password for this user, the database name and the path to the backup file have to be specified in the environment before starting the dev stack:
+```bash
+MSSQL_PASSWORD=Is1This2The3Real4Life?  # Password for the sa database user
+MSSQL_DB=MyDatabase  # Name of the database to which backup is imported
````
Können wir diese Parameter für's Devsetup loswerden und hardcoden? Wirkt auf mich wie unnötige Konfiguration.

slavistan replied:
Nvm - Production Umgebung gleichen wir mit Dev Umgebung an.

slavistan commented on README.md:
```diff
@@ -0,0 +1,47 @@
+# Development setup
+Please make sure you have a proper *python* setup with version >= 3.12.
+Another prerequisite is *poetry* since we use it for dependency management.
+If it's not already installed on your system, have a look at https://python-poetry.org/docs/.
```
Würdest du bitte

1. zwischen den Überschriften/Paragraphen/Codeblocks und dem nachfolgenden Inhalt eine Leerzeile lassen. Es wirkt sehr gedrungen.
2. Paragraphen ohne Zeilenumbrüche formattieren. Zeilenumbrüche machen den Markdownquelltext unleserlich, wenn die Breite des Fensters nicht zufällig passend eingestellt ist.

slavistan commented on README.md:
````diff
@@ -0,0 +1,47 @@
+For now, the database used within the development environment uses the default mssql port (1433) and the default mssql user (sa).
+The password for this user, the database name and the path to the backup file have to be specified in the environment before starting the dev stack:
+```bash
+MSSQL_PASSWORD=Is1This2The3Real4Life?  # Password for the sa database user
+MSSQL_DB=MyDatabase  # Name of the database to which backup is imported
+DB_BACKUP_FILE=MyDatabaseBackup.bacpac  # path to database backup
````
Sollte noch ein `export ` hin, sonst kopieren das Leute in ihre Shell und nichts passiert :(

slavistan commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+      retries: 3
+      start_period: 15s
+  db-init:
+    build:
+      context: .
+      dockerfile: ./Dockerfile
```
Können wir den Images und Containern identifizierbare Namen geben? Etwa nmshd-bkb-data-dashboard-dev-db{-init}? Mein `docker {container,image} ls` sieht richtig scheisse aus und ich würde es gerne besser machen. Geht das aus dem Kontext eines docker-compose.yml heraus?

slavistan commented on docker/development/load_bacpac.sh:
```diff
@@ -0,0 +1,29 @@
+# if database exists, exit
+if sqlcmd -S $MSSQL_HOST,$MSSQL_PORT -U $MSSQL_USER -P $MSSQL_PASSWORD -C -Q "SELECT name FROM sys.databases WHERE name = '$MSSQL_DB'" | grep -q $MSSQL_DB; then
+    echo "Database $MSSQL_DB already exists. Please drop it before running $0 or start hacking."
+    exit 1
```
Returncode 0, "database already bootstrapped"

slavistan commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+        condition: service_healthy
+    command: /db-init/load_bacpac.sh
+    environment:
+      - MSSQL_HOST=db
```
Wenn möglich, Warnung zu fehlenden Umgebungsvariablen zu einem Fehler machen.

slavistan commented on docker/development/load_bacpac.sh:
```diff
@@ -0,0 +1,29 @@
+#! /usr/bin/env bash
```
Leerzeichen im Shebang.

slavistan commented on docker/development/load_bacpac.sh:
```diff
@@ -0,0 +1,29 @@
+#! /usr/bin/env bash
+
+# Make sure sqlpackage is installed
+if ! command -v sqlpackage &> /dev/null; then
```
Existenz von sqlcmd auch prüfen, oder KISS beides rauslassen.

slavistan commented on docker/production/docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+services:
+  webserver:
+    container_name: nmshdBkbDataDashboard
```
Namen prüfen.

slavistan commented on docker/production/docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+    container_name: nmshdBkbDataDashboard
+    image: ghcr.io/js-soft/nmshd-bkb-data-dashboard:${DASHBOARD_VERSION}
+    ports:
+      - "5000:5000/tcp"
```
Hostport parametrisieren `${DASHBOARD_PORT:-5000}`.

slavistan commented on docker/production/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+gunicorn --workers $GUNICORN_WORKERS --bind $SERVER_HOST:5000 "main:create_app()"
+EOT
+ENTRYPOINT [ "/entrypoint.sh" ]
```
`ENTRYPOINT ["/bin/sh", "-c", "gunicorn --workers \$GUNICORN_WORKERS --bind \$SERVER_HOST:5000 \"main:create_app()\""]` mal ausprobieren.

slavistan commented on docker/development/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+      ACCEPT_EULA: "Y"
+      SA_PASSWORD: ${MSSQL_PASSWORD}
+    healthcheck:
+      test: /opt/mssql-tools18/bin/sqlcmd -S localhost -U sa -P "${MSSQL_PASSWORD}" -C -Q "SELECT 1" -b -o /dev/null
```
User `sa` ist hardgecoded, verträgt sich nicht mit der Konfiguration über `MSSQL_USER`. Können wir das überhaupt konfigurierbar machen? Wenn ich mich recht entsinne, liefert das Dockerimage des MSSQL-Servers den Standardnutzer 'sa' mit aus. Wenn wir den nicht nutzen, müssten wir ja vorher einen anderen anlegen, richtig?

slavistan replied:
localhost auch hardgecoded.

Philipp36 replied:
> localhost auch hardgecoded.

`localhost` ist hardgecoded, weil wir den healthcheck auf dem Container, auf dem der Server läuft, selbst durchführen. Daher können wir hier davon ausgehen, dass der Host sich nicht ändert.

slavistan replied:
Ja, war nonsense von mir. Die Datenbank für's Devsetup soll ja lokal laufen. Alles gut 👍

Philipp36 replied:
> User `sa` ist hardgecoded, verträgt sich nicht mit der Konfiguration über `MSSQL_USER`. Können wir das überhaupt konfigurierbar machen? Wenn ich mich recht entsinne, liefert das Dockerimage des MSSQL-Servers den Standardnutzer 'sa' mit aus. Wenn wir den nicht nutzen, müssten wir ja vorher einen anderen anlegen, richtig?

Genau, das habe ich der Einfachheit halber daher weggelassen und dementsprechend auch im anderen Container des Serivce die Parameter, die wir **nicht** ändern können, bereits belegt. Die Frage ist, ob wir uns noch die Mühe machen und es anpassen oder wir es zum aktuellen Stand erstmal so lassen.

Philipp36 replied:
> Ja, war nonsense von mir. Die Datenbank für's Devsetup soll ja lokal laufen. Alles gut 👍

Das meine ich nicht mal. Wenn ich das richtig verstehe, wird der Healthcheck auf dem Container selbst ausgeführt, daher ist `localhost` so oder so okay. So verstehe ich das jedenfalls, die Doku ist dort etwas unklar, aber `docker inspect --format "{{json.State.Health }}"` lässt darauf schließen.

slavistan commented on docker/production/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+# Set environment variables
+ARG GUNICORN_VERSION="23.0.0"
+ENV GUNICORN_WORKERS=4
+ENV SERVER_HOST=0.0.0.0
```
Verwirrender Name (wir haben lauter Server überall). Brauchen wir die Variable überhaupt oder kannste das hardcoden?

slavistan commented on docker/production/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+RUN poetry install --no-root && rm -rf $POETRY_CACHE_DIR
+
+FROM python:3.12-slim-bookworm AS runner
+# Install curl, gpg for healthcheck and odbc download
+RUN apt-get update && apt-get install -y \
+curl \
+gnupg2 \
+&& rm -rf /var/cache/apt/archives /var/lib/apt/lists/*
```
Nitpick: Kommentar ist irgendwie unnötig, jedes Dockerfile beginnt mit apt-get update etc :p

Und sei so lieb und rück die Multilinestatments ein.

Philipp36 committed:
WIP: PR feedback

Philipp36 committed:
WIP: docker pipeline test

Philipp36 committed:
WIP: docker pipeline

Philipp36 committed:
WIP: docker pipeline test

Philipp36 committed:
WIP: PR feedback

- add test for `sqlcmd` availability
- Exitcode 0 (indicating no eror) if database already bootstrapped for
  dev setup
- Port of MSSQL server configurable by env variable in dev setup
- Renaming of generic `WEBSERVER_*` vars
- Refactoring of Dockerfiles

slavistan committed:
rewrite readme

Philipp36 committed:
PR feedback

Philipp36 committed:
PR feedback

slavistan committed:
update readme

Philipp36 committed:
Fix: Use correct connection object

Philipp36 committed:
Chore: Update dependencies

- removed unused dependencies
- moved dependencies to "dev" group
- only install main dependencies in docker image

Philipp36 committed:
Merge branch 'feature/docker' of github:js-soft/nmshd-bkb-data-dashboard into feature/docker

Philipp36 committed:
Refactor: Renaming of db service to mssql-server in dev setup

slavistan commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+    && ACCEPT_EULA=y apt-get install -y --no-install-recommends msodbcsql18 \
+    && rm -rf /var/cache/apt/archives /var/lib/apt/lists/*
+
+# Set environment variables
```
Nitpick: Unnötiger Kommentar.

slavistan commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+RUN /app/.venv/bin/pip install gunicorn==23.0.0
+
+# Note: We use exec to replace the shell process with the gunicorn process to receive UNIX signals (and use the shell for environment variable processing)
+ENTRYPOINT [ "/bin/sh", "-c", "exec /app/.venv/bin/python -m gunicorn --workers ${DASHBOARD_NUM_WORKERS} --bind 0.0.0.0:5000 'main:create_app()'"]
```
Leerzeichen hinter der Klammer

slavistan commented on dev/bootstrap-mssql.Dockerfile:
```diff
@@ -0,0 +1,36 @@
+RUN apt-get update && ACCEPT_EULA=Y apt-get install -y --no-install-recommends \
+        mssql-tools18 \
+        unixodbc-dev \
+        && rm -rf /var/lib/apt/lists/*
```
Ein mal zu viel eingerückt :D

slavistan commented on dev/bootstrap-mssql.Dockerfile:
```diff
@@ -0,0 +1,36 @@
+        && rm -rf /var/lib/apt/lists/*
+
+# Make executables of sqltools available in PATH
+ENV PATH="$PATH:/opt/mssql-tools18/bin"
```
Genau andersrum: `PATH="/opt/mssql-tools18/bin:$PATH"`, sonst haben deine Binaries die geringste Priorität. Macht in nem nackten Dockerimage nichts, führt in anderem Kontext aber zu Fehlern und "riecht" daher komisch.

slavistan replied:
Nur aus Interesse: Warum installiert apt (bzw. die Maintainer des Pakets) Nutzerbinaries (mssql-tools18) an einen Ort, der nicht im PATH liegt? Ist das so dokumentiert?

slavistan commented on dev/bootstrap-mssql.docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+      timeout: 3s
+      retries: 3
+      start_period: 15s
+  mssql-server-init:
```
Nitpick: Der Service initialisiert wiederum nicht den mssql-Server, sondern erzeugt und befüllt eine Datenbank innerhalb des Servers. Besser also `mssql-server-db-{init,bootstrap}`.

slavistan commented on dev/bootstrap-mssql.docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+        condition: service_healthy
+    command: /db-init/load_bacpac.sh
+    environment:
+      - MSSQL_HOST=mssql-server
```
Das geht? Cool :-)

slavistan commented on dev/bootstrap-mssql.docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+      - MSSQL_HOST=mssql-server
+      - MSSQL_PORT=1433
+      - MSSQL_USER=sa
+      - MSSQL_PASSWORD=${MSSQL_PASSWORD?}
```
Lass das Passwort für den Devbetrieb doch auch gleich hardcoden.

slavistan replied:
Bohemian_Rhapsody2024

slavistan committed:
update readme

Philipp36 committed:
PR feedback

Philipp36 committed:
Merge branch 'feature/docker' of github:js-soft/nmshd-bkb-data-dashboard into feature/docker

Philipp36 committed:
MSSQL_HOST -> MSSQL_HOSTNAME

Philipp36 committed:
PR feedback
